components/merkledb/src/views/address.rs

```diff
@@ -0,0 +1,207 @@
+use std::{borrow::Cow, num::NonZeroU64};
+
+use crate::BinaryKey;
+
+pub fn key_bytes<K: BinaryKey + ?Sized>(key: &K) -> Vec<u8> {
+    concat_keys!(key)
+}
+
+/// Represents the address of an index in the database.
+///
+/// An address has a string *name* and an optional byte *key*. An index is uniquely identified
+/// by its address. Different addresses correspond to different indexes. Addresses with the same
+/// name and differing prefixes are said to belong to the same *group* (see also [`Group`]). Groups
+/// can be used for a potentially unbounded group of indexes that can be identified by a certain
+/// key (for example, `ProofListIndex` with the transaction history of a wallet keyed by the
+/// `PublicKey` of the wallet).
+///
+/// [`Group`]: struct.Group.html
+///
+/// # Examples
+///
+/// `IndexAddress` can be used implicitly, since `&str` and `(&str, &impl BinaryKey)` can both
+/// be converted into an address.
+///
+/// ```
+/// use exonum_merkledb::{access::AccessExt, IndexAddress, TemporaryDB, Database};
+///
+/// let db = TemporaryDB::new();
+/// let fork = db.fork();
+///
+/// // Using a string address:
+/// let map = fork.get_map::<_, String, u8>("map");
+/// // Using an address within an index family:
+/// let list = fork.get_list::<_, String>(("index", &3_u32));
+/// // Using `IndexAddress` explicitly:
+/// let addr = IndexAddress::from_root("data").append_key(&vec![1, 2, 3]);
+/// let set = fork.get_value_set::<_, u64>(addr);
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Default)]
+pub struct IndexAddress {
+    pub(super) name: String,
+    pub(super) id_in_group: Option<Vec<u8>>,
+}
+
+impl IndexAddress {
+    /// Creates new `IndexAddress` with the specified name.
+    pub fn from_root<S: Into<String>>(root: S) -> Self {
+        Self {
+            name: root.into(),
+            id_in_group: None,
+        }
+    }
+
+    /// Returns the name part of `IndexAddress`.
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// Returns the bytes part of `IndexAddress`.
+    pub fn id_in_group(&self) -> Option<&[u8]> {
+        self.id_in_group.as_ref().map(Vec::as_slice)
+    }
+
+    /// Prepends a name part to `IndexAddress`. The name is separated from the existing name
+    /// by a dot `.`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use exonum_merkledb::IndexAddress;
+    /// let addr = IndexAddress::from_root("foo");
+    /// let prefixed = addr.prepend_name("prefix");
+    /// assert_eq!(prefixed.name(), "prefix.foo");
+    /// ```
+    pub fn prepend_name(self, prefix: &str) -> Self {
+        let name = if self.name.is_empty() {
+            prefix.to_owned()
+        } else {
+            // Because `concat` is faster than `format!("...")` in all cases.
+            [prefix, ".", self.name()].concat()
```

Comment:
Can we use [`std::concat`](https://doc.rust-lang.org/std/macro.concat.html) here? Then I think it will look more natural and the comment will not be required anymore.

Reply:
`concat!` concatenates literals, not values; it cannot be used with variables.

Reply:
Ah, I see.